### PR TITLE
Allowed services support

### PR DESCRIPTION
**Pull request by zakorgy**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

### CODE REVIEW

**Comment by zakorgy** [comment]
@dati91  r?

File: components/net/bluetooth_thread.rs [hunk 1]
```diff
@@ -139,6 +140,7 @@ pub struct BluetoothManager {
     cached_services: HashMap<String, BluetoothGATTService>,
     cached_characteristics: HashMap<String, BluetoothGATTCharacteristic>,
     cached_descriptors: HashMap<String, BluetoothGATTDescriptor>,
+    allowed_services: HashMap<String, HashSet<String>>, // device ids to list of UUIDs
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: `// device id to UUIDs`

File: components/net_traits/bluetooth_scanfilter.rs [hunk 2]
```diff
@@ -90,4 +94,8 @@ impl RequestDeviceoptions {
     pub fn get_filters(&self) -> &BluetoothScanfilterSequence {
         &self.filters
     }
+
+    pub fn get_optional_services(&self) -> &ServiceUUIDSequence {
+        (&self.optional_services).clone()
```
Comment (dati91) [inline components/net_traits/bluetooth_scanfilter.rs]: we really need this `clone`?

File: components/net/bluetooth_thread.rs [hunk 3]
```diff
@@ -522,6 +539,11 @@ impl BluetoothManager {
                             uuid: Option<String>,
                             sender: IpcSender<BluetoothResult<BluetoothServicesMsg>>) {
         let mut adapter = get_adapter_or_return_error!(self, sender);
+        if let Some(ref uuid) = uuid {
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: with this line, we duplicate the `match uuid` logic

Comment (dati91) [inline components/net/bluetooth_thread.rs]: but the code is cleaner this way i'll give you that

File: components/net/bluetooth_thread.rs [hunk 4]
```diff
@@ -139,6 +140,7 @@ pub struct BluetoothManager {
     cached_services: HashMap<String, BluetoothGATTService>,
     cached_characteristics: HashMap<String, BluetoothGATTCharacteristic>,
     cached_descriptors: HashMap<String, BluetoothGATTDescriptor>,
+    allowed_services: HashMap<String, HashSet<String>>, // device id to UUIDs
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: maybe put the comment above this code

File: components/net/bluetooth_thread.rs [hunk 5]
```diff
@@ -450,16 +451,10 @@ impl BluetoothManager {
...
-                   .iter()
-                   .map(|filter| filter.get_services().iter().map(|s| services.insert(s.clone())).collect::<Vec<_>>())
-                   .collect::<Vec<_>>();
-            options.get_optional_services()
-                   .get_services()
-                   .iter()
-                   .map(|s| services.insert(s.clone()))
-                   .collect::<Vec<_>>();
+            let mut services = options.get_services_set();
+            if let Some(servives_set) = self.allowed_services.get(&device_id) {
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: maybe `get_mut` and `services_set |= services;`
then we don't need the `insert` line

**Comment by dati91** [comment]
lgtm

**Pull request merged**